server/src/main/java/org/opensearch/cluster/store/RemoteClusterStateService.java

```diff
@@ -0,0 +1,247 @@
+        this.repositoriesService = repositoriesService;
+        this.settings = settings;
+    }
+
+    @Nullable
+    public ClusterMetadataMarker writeFullMetadata(ClusterState clusterState) throws IOException {
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        initializeRepository();
+        if (blobStoreRepository == null) {
+            logger.error("Unable to set repository");
+            return null;
+        }
+
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = new HashMap<>();
+        // todo parallel upload
+        // any validations before/after upload ?
+        for (IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
+            // 123456789012_test-cluster/cluster-state/dsgYj10Nkso7/index/ftqsCnn9TgOX/metadata_4_1690947200
+            final String indexMetadataKey = writeIndexMetadata(
+                clusterState.getClusterName().value(),
+                clusterState.getMetadata().clusterUUID(),
+                indexMetadata,
+                indexMetadataFileName(indexMetadata)
+            );
+            final UploadedIndexMetadata uploadedIndexMetadata = new UploadedIndexMetadata(
+                indexMetadata.getIndex().getName(),
+                indexMetadata.getIndexUUID(),
+                indexMetadataKey
+            );
+            allUploadedIndexMetadata.put(indexMetadata.getIndex().getName(), uploadedIndexMetadata);
+        }
+        return uploadMarker(clusterState, allUploadedIndexMetadata);
+    }
+
+    @Nullable
+    public ClusterMetadataMarker writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataMarker previousMarker
+    ) throws IOException {
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        final Map<String, Long> indexMetadataVersionByName = new HashMap<>();
+        for (final IndexMetadata indexMetadata : previousClusterState.metadata().indices().values()) {
+            indexMetadataVersionByName.putIfAbsent(indexMetadata.getIndex().getName(), indexMetadata.getVersion());
+        }
+
+        int numIndicesUpdated = 0;
+        int numIndicesUnchanged = 0;
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = new HashMap<>(
+            previousMarker.getIndices()
+        );
+        for (final IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
```

Comment:
Is it possible that `previousMarker` does not contain all the indices from `previousClusterState`? If yes, we need to upload metadata of those indices that are still part of clusterState.